### PR TITLE
Preserve parens required by C# when simplifying interpolation

### LINKED ISSUE #49647: 'Simplify interpolation' fix should not remove parens around a conditional expression
**Version Used**: VS 16.8.2 and VS 16.9-p1

(How soon is too soon to start reporting bugs assuming top level statements are turned on for new tests by default so I can exclude `class C` etc?)

```cs
class C
{
    void M(bool cond)
    {
        // ℹ IDE0071 Interpolation can be simplified
        //     ↓
        _ = $"{(cond ? 1 : 2).ToString()}";
    }
}
```

💡 Simplify interpolation:

```cs
class C
{
    void M(bool cond)
    {
        // ❌ CS8361 A conditional expression cannot be used directly in a string interpolation because the ':' ends the
        // interpolation. Parenthesize the conditional expression.
        //     ↓↓↓↓↓↓↓↓↓
        _ = $"{cond ? 1 : 2}";
    }
}
```

Expected:

```cs
class C
{
    void M(bool cond)
    {
        _ = $"{(cond ? 1 : 2)}";
    }
}
```

### PR DESCRIPTION
**Pull request by jnm2**
Fixes #49647

### CODE REVIEW

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Analyzers/Core/Analyzers/SimplifyInterpolation/Helpers.cs [hunk 1]
```diff
@@ -14,9 +14,14 @@
...
-    internal static class Helpers
+    internal class Helpers
     {
-        public static void UnwrapInterpolation<TInterpolationSyntax, TExpressionSyntax>(
+        protected virtual SyntaxNode GetPreservedInvocationInstanceSyntax(IInvocationOperation invocationOperation)
+        {
+            return invocationOperation.Instance.Syntax;
+        }
+
+        public void UnwrapInterpolation<TInterpolationSyntax, TExpressionSyntax>(
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/Analyzers/SimplifyInterpolation/Helpers.cs:24]: alternative approach.  update this helper to have TConditional/TParenthesized.  this helper can then check the syntax shape and adjust right here.

Comment (jnm2) [reply src/Analyzers/Core/Analyzers/SimplifyInterpolation/Helpers.cs:24]: This is a very C#-specific thing to be checking. What would the parameters be for VB?

Comment (jnm2) [reply src/Analyzers/Core/Analyzers/SimplifyInterpolation/Helpers.cs:24]: Just checked, VB never removes parens anyway, so there is no downside to hardcoding this across all languages.
https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBsQDdhJhxBLAMwE8AfASTAAcB7BGKAAgGUjY4wBYAKAAUBXYCjwBjBgGEU0RmO4M5DfoJEMAYnwB2wmHmrqGAWQAUASgYBBRkxgI86gOaz5TgEpwYfBHoAkAIgDehgBMxgB0ACrUVjb2JgC+Po5yAKLqACaqGlo66twp6RJSQA

File: src/Analyzers/Core/Analyzers/SimplifyInterpolation/Helpers.cs [hunk 2]
```diff
@@ -14,9 +14,14 @@
...
 namespace Microsoft.CodeAnalysis.SimplifyInterpolation
 {
-    internal static class Helpers
+    internal class Helpers
     {
-        public static void UnwrapInterpolation<TInterpolationSyntax, TExpressionSyntax>(
+        protected virtual SyntaxNode GetPreservedInvocationInstanceSyntax(IInvocationOperation invocationOperation)
+        {
+            return invocationOperation.Instance.Syntax;
+        }
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/Analyzers/SimplifyInterpolation/Helpers.cs:22]: far prefer this to be abstract.  make bth subclasses be explicit about the behavior they want.

File: src/Analyzers/Core/Analyzers/SimplifyInterpolation/AbstractSimplifyInterpolationDiagnosticAnalyzer.cs [hunk 3]
```diff
@@ -32,6 +32,8 @@ protected AbstractSimplifyInterpolationDiagnosticAnalyzer()
 
         protected abstract ISyntaxFacts GetSyntaxFacts();
 
+        protected virtual Helpers GetHelpers() => new();
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/Analyzers/SimplifyInterpolation/AbstractSimplifyInterpolationDiagnosticAnalyzer.cs:35]: make this abstract as well.

File: src/Analyzers/Core/Analyzers/SimplifyInterpolation/Helpers.cs (hunk 1 above)
Comment (jnm2) [reply src/Analyzers/Core/Analyzers/SimplifyInterpolation/Helpers.cs]: Implemented.

File: src/Analyzers/Core/Analyzers/SimplifyInterpolation/Helpers.cs (hunk 2 above)
Comment (jnm2) [reply src/Analyzers/Core/Analyzers/SimplifyInterpolation/Helpers.cs:29]: (Inheritance was removed.)

File: src/Analyzers/Core/Analyzers/SimplifyInterpolation/AbstractSimplifyInterpolationDiagnosticAnalyzer.cs (hunk 3 above)
Comment (jnm2) [reply src/Analyzers/Core/Analyzers/SimplifyInterpolation/AbstractSimplifyInterpolationDiagnosticAnalyzer.cs]: (Inheritance was removed.)

**Comment by CyrusNajmabadi** [comment]
Thanks!

**Comment by jnm2** [comment]
Thanks for your help!

**Pull request merged**